modules/stochastic_tools/src/surrogates/GaussianProcessTrainer.C

```diff
@@ -46,16 +61,53 @@ GaussianProcessTrainer::GaussianProcessTrainer(const InputParameters & parameter
     _data_standardizer(declareModelData<StochasticTools::Standardizer>("_data_standardizer")),
     _K(declareModelData<RealEigenMatrix>("_K")),
     _K_results_solve(declareModelData<RealEigenMatrix>("_K_results_solve")),
+    _K_cho_decomp(declareModelData<Eigen::LLT<RealEigenMatrix>>("_K_cho_decomp")),
     _standardize_params(getParam<bool>("standardize_params")),
     _standardize_data(getParam<bool>("standardize_data")),
     _covar_type(declareModelData<std::string>("_covar_type")),
     _hyperparam_map(declareModelData<std::unordered_map<std::string, Real>>("_hyperparam_map")),
     _hyperparam_vec_map(declareModelData<std::unordered_map<std::string, std::vector<Real>>>(
         "_hyperparam_vec_map")),
     _covariance_function(
-        getCovarianceFunctionByName(getParam<UserObjectName>("covariance_function")))
+        getCovarianceFunctionByName(getParam<UserObjectName>("covariance_function"))),
+    _do_tuning(getParam<bool>("tune")),
+    _tao_options(getParam<std::string>("tao_options")),
+    _show_tao(getParam<bool>("show_tao"))
 
 {
+#ifndef LIBMESH_HAVE_PETSC
+  if (_do_tuning)
+    ::mooseError("Hyperparameter tuning requires PETSc/TAO");
+#endif
+
+  _num_tunable = 0;
+  std::vector<std::string> tune_parameters(getParam<std::vector<std::string>>("tune_parameters"));
+  // Error Checking
+  if (isParamValid("tuning_min") &&
+      (getParam<std::vector<Real>>("tuning_min").size() != tune_parameters.size()))
+    ::mooseError("tuning_min size does not match tune_parameters");
+  if (isParamValid("tuning_max") &&
+      (getParam<std::vector<Real>>("tuning_max").size() != tune_parameters.size()))
+    ::mooseError("tuning_max size does not match tune_parameters");
+  // Fill Out Tunable Paramater information
+  for (unsigned int ii = 0; ii < tune_parameters.size(); ++ii)
+  {
+    const auto & hp = tune_parameters[ii];
+    if (_covariance_function->isTunable(hp))
```

Comment:
See other response. At the moment I still have the error in the original spot. Basically there can be two reasons why a parameter is not tunable. It may be a valid parameter in the list, however tuning is not enables for that parameter.

However, it could just be a nonsense/mistyped name.

Regardless, it should return False.

I can rework this and pull the isParamValid out of isTunable, but I am not sure if that is any better.

Reply:
I agree with you now. I was not considering the fact that a parameter might be valid, but not tunable.